### PR TITLE
adding tutorial for creating a Moore's law linear regression

### PR DESCRIPTION
**Pull request by cooperrc**
This tutorial builds a linear regression model based upon historical semiconductor manufacturing data for number of transistors per microchip. It teaches a a typical user workflow by:

1. You'll learn to load data from a comma separated values (*.csv) file
2. You'll use ordinary least squares to do linear regression and predict exponential growth
3. You'll compare exponential growth constants between models
4. You'll share your analysis as NumPy zipped files (*.npz)
5. You'll share your analysis as a *.csv file
6. You'll assess the amazing progress semiconductor manufacturers have made in the last five decades

Some notes on this PR:
1. I originally started writing this to address issue [#17266](https://github.com/numpy/numpy/issues/17266#issue-695390569), but the scope of building a linear regression model made it more difficult to directly address this issue. I can write a seperate, shorter tutorial to save and load data. 
2. ~~I do not know how to build intersphinx references for the functions I used. In the notebook I made html comments 
  `<!--need ... reference-->` 
  in each spot where I thought there should be Sphinx link.~~ @mattip said our best-practice is numpy.org/doc/stable<links> so I updated the references 
3. What makes this tutorial great is that it has real data and compares the results to Gordon Moore's prediction. It keeps the reader focused on learning, then building, comparing, then finally sharing the results. 

Thanks to @melissawm, @mattip, @bjnath, and @rgommers for guidance and initial reviews.

### CODE REVIEW

**Comment by melissawm** [comment]
I think the failure might be related to the outputs from the notebook. Can you try removing all outputs before submitting and see if that helps? Thanks!

**Comment by cooperrc** [comment]
> I think the failure might be related to the outputs from the notebook. Can you try removing all outputs before submitting and see if that helps? Thanks!

@melissawm, I ran [`clean_ipynb`](https://github.com/KwatME/clean_ipynb) to remove the output from the notebook. I didn't realize that we shouldn't have output in the notebook cells. There are less errors now, but we will need to add `statsmodels` the the repo dependencies.

**Comment by bjnath** [comment]
Clearly a lot of thought and effort went into this!  I have just a few comments on the opening section.

Regarding this sentence

> the **rate** of **semiconductor**s on a **computer** chip 

* Did you mean to say it's the **number** (rather than the **rate**) that will double?
* Better to say it's transistor count that's doubling ("semiconductor" is the chip material).
* Although we're focusing the analysis on microprocessors, they're a subset of integrated circuits (there's also memory chips, for instance); I think Moore's prediction was for all ICs. It's more interesting to look at microprocessors because they emerge more often than new generations of memory. (Memory chips should follow a similar curve but with higher transistor counts -- memories are denser because of a regular layout.)

---
I'd propose this opening:

**What you'll do**

In 1965, engineer Gordon Moore [predicted](https://en.wikipedia.org/wiki/Moore's_law) that transistors on a chip would double every two years in the coming decade. We'll compare that against actual transistor counts in the 53 years afterward.

**Skills you'll learn**

* Load data from a [*.csv](https://en.wikipedia.org/wiki/Comma-separated_values) file
* Perform linear regression and predict exponential growth using ordinary least squares
* Share your analysis in a file:
  *  a NumPy zipped file (`*.npz`)
  * a `*.csv` file

**What you'll need**

1. These packages
 * NumPy
 * Matplotlib
 * statsmodels ordinary linear regression

  imported with the commands
 ```
 import numpy as np
 import matplotlib.pyplot as plt
 import statsmodels.api as sm
```
We'll be using use these libraries from those packages:
...

2. Since this is an exponential law, you'll need a little background in natural logs and exponentials.

---

I don't think I allowed for this in my tutorial guidelines, but since there are many steps it might be good to summarize them first -- possibly just with a table of contents.

**Comment by cooperrc** [comment]
@bjnath, I really like the suggested edits. Thanks! I can update the tutorial. 

@melissawm, I think the failed checks now stem from the missing `statsmodels`. 

```
----> 3 import statsmodels.api as sm

ModuleNotFoundError: No module named 'statsmodels'
```
It looks like the subsequent errors come from missing `sm.<func>` assignments.

**Comment by melissawm** [comment]
> @bjnath, I really like the suggested edits. Thanks! I can update the tutorial.
> 
> @melissawm, I think the failed checks now stem from the missing `statsmodels`.
> 
> ```
> ----> 3 import statsmodels.api as sm
> 
> ModuleNotFoundError: No module named 'statsmodels'
> ```
> 
> It looks like the subsequent errors come from missing `sm.<func>` assignments.

I just opened #32 to fix that, should work. Thanks @bjnath for the suggestions, @cooperrc you can update the notebook and as soon as #32 is merged we should have no other problems.

**Comment by bjnath** [comment]
Would it be less scary to start from the equation itself? (Provided I've done it correctly...)

**Building Moore's law**

Making "doubles every two years" into an equation, if the transistor count in year *Y* is *T*, in year *n* it will be

<a href="https://www.codecogs.com/eqnedit.php?latex=t&space;=&space;T&space;*&space;2^{(n-Y)/2}" target="_blank"></a>

That is, two years later *T* is multiplied by a factor of 

<a href="https://www.codecogs.com/eqnedit.php?latex=2^{(T&plus;2-T)/2}&space;=&space;2^{2/2}&space;=&space;2^1&space;=&space;2" target="_blank"></a>

in four years, by 

<a href="https://www.codecogs.com/eqnedit.php?latex=2^{(T&plus;4-T)/2}&space;=&space;2^{4/2}&space;=&space;2^2&space;=&space;2&space;*&space;2" target="_blank"></a>

and so on.

If we plot this, it's an exponential curve:
(possibly a plot here)

But we can make it a flat line by taking the log of both sides:


<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align}&space;log_2(t)&space;&=&space;log_2(T*&space;2^{(n-Y)/2})&space;\notag&space;\\&space;&=&space;log_2(T)&space;&plus;&space;log_2(2^{(n-Y)/2})&space;\notag&space;\\&space;&=&space;log_2(T)&space;&plus;&space;(n-Y)/2&space;\notag&space;\\&space;&=&space;log_2(T)&space;&plus;&space;n/2&space;-&space;Y/2&space;\notag&space;\\&space;\notag&space;\end{align}" target="_blank"></a>

This has the familar form *y = ax + b*, where 


<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align}&space;y&space;&=&space;log_2(t)&space;\notag&space;\\&space;a&space;&=&space;1/2&space;\notag&space;\\&space;b&space;&=&space;log_2(T)&space;-&space;Y/2&space;\notag&space;\\&space;\notag&space;\end{align}" target="_blank"></a>

There were 2250 transistors in 1971, so

Y = 1971
T = 2250

giving the straight-line formula

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align}&space;log_2(t)&space;&=&space;n/2&space;&plus;&space;log_2(2250)-1971&space;\notag&space;\\&space;&=&space;n/2&space;-&space;974.4&space;\notag&space;\end{align}" target="_blank"></a>

**Comment by cooperrc** [comment]
> Would it be clearer to start from the law itself?
> 
> **Building Moore's law**

I had considered this too, but I thought it would be worthwhile to keep the two functions as similar as possible i.e. linear on a natural log scale. 

You have to go through a few steps to get to the log2, log10, and loge, so I just kept it as loge, since that is the default NumPy/Mathematics logarithm. _In engineering, loge is ln and log is log10, so there is that extra layer of confusion._

**Comment by melissawm** [comment]
@cooperrc if you merge master now, the tests should pass. Sorry for taking so long!

**Comment by cooperrc** [comment]
> @cooperrc if you merge master now, the tests should pass. Sorry for taking so long!

No need to apologize! I took some time to remove passive voice and add the docs/stable links in the notebook. Should be much better now. Plus, it passes all checks!

**Comment by melissawm** [comment]
This looks good to me! I have only a few (final) minor comments, then I feel like this can be merged.

File: content/mooreslaw-tutorial.ipynb
Comment (melissawm) [inline content/mooreslaw-tutorial.ipynb:1]: <ul><li>Can you add the image that should be shown in the first cell of the notebook to the repo?</li><li>Is the first reference in the "What you'll do" section supposed to have a link, like the second one? Also, this paragraph has a trailing "u".</li><li>I believe there is an extra backslash in <code>*.npz</code></li><li> It would be nice to add links to the sites for matplotlib (https://matplotlib.org) and statsmodels (https://statsmodels.org)</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/numpy/numpy-tutorials/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (melissawm) [inline content/mooreslaw-tutorial.ipynb:1]: <p>In <code>np.exp</code>  I think you copied the explanation for the <code>log</code>  function</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/numpy/numpy-tutorials/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (melissawm) [inline content/mooreslaw-tutorial.ipynb:1]: <p>You determine these constants for Moore's law by specifying the rate for added <strong>transistors</strong>, 2 (missing a t in transistors)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/numpy/numpy-tutorials/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (melissawm) [inline content/mooreslaw-tutorial.ipynb:1]: <p>Extra slash in filename might be trouble for Windows users.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/numpy/numpy-tutorials/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (melissawm) [inline content/mooreslaw-tutorial.ipynb:1]: <p>I feel like these two sentences</p><p> </p><p><code>You directly compare the number of predicted transistors to the actual number of transistors per microprocessor over the course of manufacturing history. Plot the linear regression results and all of the transistor counts.</code></p><p> </p><p>may be unnecessary from the surrounding text. I think you mention in the paragraph before that you will compare the model with the actual reports, and in the next paragraph you will do the plotting. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/numpy/numpy-tutorials/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (melissawm) [inline content/mooreslaw-tutorial.ipynb:1]: <p>It might be nice to add a quick link to the <code>plt.style</code>  documentation (the other commands are fairly standard so there's no need).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/numpy/numpy-tutorials/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (melissawm) [inline content/mooreslaw-tutorial.ipynb:1]: <p>Extra slash here and in the cell below.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/numpy/numpy-tutorials/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (melissawm) [inline content/mooreslaw-tutorial.ipynb:1]: <p>If you want to share data <strong>(and?)</strong> view the results in a table - something is missing in this sentence.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/numpy/numpy-tutorials/pull/31/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

Comment (cooperrc) [reply content/mooreslaw-tutorial.ipynb:1]: Great ideas. Thanks for catching the typos
Should images be located in the "images" folder for the repo?

Comment (cooperrc) [reply content/mooreslaw-tutorial.ipynb:1]: I wouldn't put it past myself, but I think the latest commit has the correct links. _I'm copy-pasting the markdown to double-check_:
* [`np.log`](https://numpy.org/doc/stable/reference/generated/numpy.log.html): this function takes the natural log of a NumPy array
* [`np.exp`](https://numpy.org/doc/stable/reference/generated/numpy.exp.html): this function takes the natural log of a NumPy array

Comment (cooperrc) [reply content/mooreslaw-tutorial.ipynb:1]: spell-check, thanks. I misspelled _semiconduct~~e~~or_ right above too.

Comment (cooperrc) [reply content/mooreslaw-tutorial.ipynb:1]: thanks

Comment (cooperrc) [reply content/mooreslaw-tutorial.ipynb:1]: You're right, its redundant. thanks for catching that.

Comment (cooperrc) [reply content/mooreslaw-tutorial.ipynb:1]: <p>I missed this, current form doesn't make sense. I wanted simplify the use case here e.g. You want to view results in a table and share a table with other users. </p><p> </p><p>"share data <strong>and</strong> view the results" makes sense.</p>

Comment (melissawm) [reply content/mooreslaw-tutorial.ipynb:1]: <p>I think the images for tutorials should maybe live inside the content repo, just to make sure they are reachable in binder (binder will set up only the content part of the repo). Not sure if this will change if the repo workflow changes.</p>

Comment (melissawm) [reply content/mooreslaw-tutorial.ipynb:1]: <p>The links are right, but np.exp takes the exponential of a NumPy array, right? :) </p>

**Comment by melissawm** [comment]
I think those are the final comments from me.

File: content/mooreslaw-tutorial.ipynb
Comment (cooperrc) [reply content/mooreslaw-tutorial.ipynb:1]: haha, yes `exp` does not take the "natural log of a NumPy array". Your comment said exactly that, whoops.

**Comment by melissawm** [comment]
Thanks, @cooperrc ! @mattip do you want to take a look before we merge?

**Review by melissawm — APPROVED** [review]

**Comment by mattip** [comment]
As per the discussion today, merging as-is, but there may be more tweaks

**Comment by mattip** [comment]
thanks @cooperrc

**Comment by cooperrc** [comment]
> thanks @cooperrc

You're very welcome! Thanks @melissawm and @bjnath for your careful review!

**Comment by 8bitmp3** [comment]
@cooperrc @melissawm @rossbar The Markdown versions/conversions of GSoD notebooks, like the Moore's Law or Deep Learning with MNIST are effectively new PRs authored by the maintainers. So, would it be OK if we add the authors' names to `.md` files after the titles (e.g. "by @cooperrc" or "Author: @cooperrc" — similar to: https://keras.io/guides/functional_api/, https://pytorch.org/tutorials/beginner/pytorch_with_examples.html) LMKWYT 👍

**Comment by rossbar** [comment]
Thanks @8bitmp3 , an author tag is definitely a good idea. I think this can already be specified in the notebook metadata, but we'd need to check that and then verify that the theme picks that up and displays it prominently in the article. If not, we can always just add our own to the template.

**Comment by rgommers** [comment]
For the record, displaying author names used to be done in doc sections a very long time ago, and we removed it because it's usually a bad idea. There are a couple of reasons for that:

- It's not relevant info to the audience
- It's often not reflective of the development process; there are usually multiple participants. 
- It's yet another thing to decide on in the future when more edits are made, and then the later contributors aren't mentioned.

> The Markdown versions/conversions of GSoD notebooks, like the Moore's Law or Deep Learning with MNIST are effectively new PRs authored by the maintainers.

I may be missing something here, but authorship needs to be preserved across commits/reworks. If what happened is that a contributor wrote a lot of content but didn't end up in the commit log as an author, that's problematic. Even if you copy over content to a new file format, you should use `git commit --author=<...>` to preserve authorship usually. And then add more changes on top.

**Comment by rossbar** [comment]
> Even if you copy over content to a new file format, you should use git commit --author=<...> to preserve authorship usually.

This is the step we were missing - it was simply something I failed to consider when converting from .ipynb -> .md formats. Any suggestions how to fix for the existing tutorials? We could amend the commit authors but that will change the hashes.

**Comment by rgommers** [comment]
So in this case the MNIST tutorial looks fine to me. @8bitmp3 has many commits in master, and @melissawm has one for the `.ipynb` -> `.md` conversion. That looks as it should, and also https://github.com/numpy/numpy-tutorials/graphs/contributors looks fine; @cooperrc and @8bitmp3 are the top two contributors to this repo.

> We could amend the commit authors but that will change the hashes.

Yeah that's bad practice in general. Here we could still consider it, in case there's something important enough to fix. There's only 15 forks right now, so fixing those up by hand isn't too painful.

**Comment by rossbar** [comment]
Yeah, the history as a whole should be fine, the commits related to the `.ipynb` work should all be properly attributed. The problem mostly arises when looking at the history only for individual files on GitHub. For example, looking at the [history for the mooreslaw-tutorial.md](https://github.com/numpy/numpy-tutorials/commits/master/content/mooreslaw-tutorial.md) doesn't show all the work @cooperrc in the original authorship of the `.ipynb` file that this was converted from.

**Comment by rgommers** [comment]
I'm not sure that matters that much. People mostly look at number of commits or full history (e.g. `gitk --all`) I think.

Related: we need a new way at some point to acknowledge contributions on the website. The old `THANKS.txt` in the root of the main `numpy` repo makes very little sense anymore. I'm thinking we should use the bot from https://allcontributors.org/. But not on the main repo, on the website one and then make the about page use it. Could also do the team gallery that way perhaps. Looking at the html that that bot generates, it could replace what we do now.

The idea is that then we could simply have an issue there that maintainers can comment on whenever a new significant contribution is made: `@bot add @username for content contributions`.

**Comment by melissawm** [comment]
I commented on https://github.com/numpy/numpy-tutorials/issues/57 before reading this - still catching up. I did not realize this problem beforehand, sorry about that.

**Pull request merged**
